scripts/subspace_learning_lib.py

```diff
@@ -0,0 +1,243 @@
+        return partial(loglikelihood_fn, predict_fn=predict_fn)(params_pytree, *data)
+
+    def projected_logprior(params_subspace):
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        return logprior_fn(params_full)
+
+    def projected_potential(params_subspace, batch):
+        X, y = batch
+        params_full = convert_params_from_subspace_to_full(params_subspace, projection_matrix, initial_full_params)
+        params_pytree = flat_to_pytree_fn(params_full)
+        batch_loglikelihood = vmap(loglikelihood_fn, in_axes=(None, 0, 0, None))(params_pytree, X, y, predict_fn)
+        return -jnp.sum(batch_loglikelihood) - projected_logprior(params_subspace)
+
+    if "centering_value" in hyperparams and hyperparams["centering_value"] is None:
+        optimizer_params = hyperparams.pop("optimizer_params") if "optimizer_params" in hyperparams \
+            else {"lr": 1e-1, "beta_1": 0.9, "beta_2": 0.999, "epsilon": 1e-7}
+
+        train_step_partial = partial(train_step,
+                                     loss=projected_potential,
+                                     hyperparams=optimizer_params,
+                                     callback=subspace_train_callback)
+
+        sampler = build_sgd(batch_size=len(dataset["X"]), train_step=train_step_partial,
+                            dataset=(dataset["X"], dataset["y"]))
+
+        initial_subspace_params = jnp.zeros((1, d))
+        hyperparams["centering_value"] = sampler(None, num_warmup, initial_subspace_params)[0]
+        num_warmup = 0
+
+    if not hyperparams:
+        sampler = sampler_init_fn(num_warmup=num_warmup,
+                                  potential=partial(projected_potential, batch=(dataset["X"], dataset["y"])))
+        num_warmup = 0
+
+    elif "optimizer_params" in hyperparams:
+        optimizer_params = hyperparams.pop("optimizer_params") if "optimizer_params" in hyperparams \
+            else {"lr": 1e-1, "beta_1": 0.9, "beta_2": 0.999, "epsilon": 1e-7}
+        train_step_partial = partial(train_step, loss=projected_potential, hyperparams=optimizer_params,
+                                     callback=subspace_train_callback)
+
+        sampler = sampler_init_fn(**hyperparams, train_step=train_step_partial, dataset=(dataset["X"], dataset["y"]))
+        num_samples += num_warmup
+        num_warmup = 0
+    else:
+        sampler = sampler_init_fn(**hyperparams, loglikelihood=projected_loglikelihood, \
+                                  logprior=projected_logprior, data=(dataset["X"][:, None, ...], dataset["y"]))
+        num_samples += num_warmup
+
+    initial_subspace_params = hyperparams["centering_value"] if "centering_value" in hyperparams else jnp.zeros((1, d))
+    sample_key, key = split(key)
+    sampler_output = sampler(sample_key, num_samples, initial_subspace_params)
+
+    final_subspace_params = tree_map(lambda x: x[num_warmup:].mean(axis=0), sampler_output)
```

Comment:
the sampler should return all the samples (in the full space), not just their mean